fossid-webapp-client/src/main/kotlin/FossIdRestService.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2020 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+@file:Suppress("TooManyFunctions")
+
+package org.ossreviewtoolkit.fossid
+
+import okhttp3.ResponseBody
+
+import org.ossreviewtoolkit.fossid.api.Project
+import org.ossreviewtoolkit.fossid.api.identification.identifiedFiles.IdentifiedFile
+import org.ossreviewtoolkit.fossid.api.result.FossIdScanResult
+import org.ossreviewtoolkit.fossid.api.status.DownloadStatus
+import org.ossreviewtoolkit.fossid.api.status.ScanStatus
+
+import retrofit2.Call
+import retrofit2.http.Body
+import retrofit2.http.GET
+import retrofit2.http.POST
+
+internal interface FossIdRestService {
+    @POST("api.php")
+    fun getProject(@Body body: PostRequestBody): Call<EntityPostResponseBody<Project>>
+
+    @POST("api.php")
+    fun getScansForProject(@Body body: PostRequestBody): Call<EntityPostResponseBody<Any>>
+
+    @POST("api.php")
+    fun createProject(@Body body: PostRequestBody): Call<MapResponseBody<String>>
```

Comment:
From the perspective of a user of this API it's not so nice that all these functions take a `PostRequestBody`, and although you already call e.g. `createProject()` you still need to also pass "create" as the action string. Couldn't you solve this by changing the signature to something like

    @POST("api.php")
    fun createScan(user: String, apiKey: String, @Body body: PostRequestBody = PostRequestBody("create", "project", apiKey, user)): Call<MapResponseBody<String>>

But this also is not yet as nice as it probably could be, since you need to still pass the `user` and `apiKey` for every call. So how about solving that bit by providing a `create()` function as part of a `companion object` in this `interface` that creates a Retrofit instance based on the passed OkHttp client but with an interceptor added that amend the post body with the `user` and `apiKey`? Something like this:

    companion object {
        fun create(url: String, user: String, apiKey: String, client: OkHttpClient? = null): FossIdRestService {
            val clientWithInterceptor = client?.newBuilder()?.addInterceptor { chain ->
                val request = chain.request()
                val requestBody = request.body
                if (requestBody != null) {
                    // Add user and apiKey to requestBody. 
                    chain.proceed(request.newBuilder().post(requestBody).build())
                } else {
                    chain.proceed(request)
                }
            }?.build()

            val retrofit = Retrofit.Builder()
                .apply { if (client != null) client(clientWithInterceptor) }
                .baseUrl(url)
                .build()

            return retrofit.create(FossIdRestService::class.java)
        }
    }

Reply:
That said, I'm actually very surprised that FOSSID expected the `user` and `apiKey` as part of the POST body. Is there no way to authenticate via a Basic auth header or so?

Reply:
> From the perspective of a user of this API it's not so nice that all these functions take a `PostRequestBody`, and although you already call e.g. `createProject()` you still need to also pass "create" as the action string. Couldn't you solve this by changing the signature to something like

As I said in a previous comment I think there is some confusion here : the class **FossIdRestService** is internal and is not meant to be used directly by developers who want to leverage our client.
In a following PR, I will introduce a public class **FossIdRestClient** which creates the retrofit client from the service.
This **FossIdRestClient** has some functions that abstract the low level API (these PostResponseBody classes) e.g.
```
fun createScan(projectCode: String, gitRepoUrl: String, gitBranch: String): String {
        **[snip]**
        val body = PostRequestBody("create", SCAN_GROUP, apiKey, user)
        [snip]
        val response = service.createScan(body).execute().body()
```
Therefore the user will never have to create a **PostRequestBody** by themselves nor pass the "_create_" magic value.
Nevertheless I find the idea with the interceptor for user/apiKey interesting and will use it in my next PR (containing the **FossIdRestClient**).

> That said, I'm actually very surprised that FOSSID expected the `user` and `apiKey` as part of the POST body. Is there no way to authenticate via a Basic auth header or so?

No basic auth is not working and only credentials in the POST body seems to work.